File: src/ai_driver.py
#!/usr/bin/env python3

import math

import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient

from geometry_msgs.msg import PoseStamped, TwistStamped
from nav2_msgs.action import NavigateToPose
from visualization_msgs.msg import Marker

from override_sim.msg import ActionState
from override_sim.srv import IntakeElement, ScoreElement


class AIDriver(Node):
    """Executes the Strategy AI output: drives to the requested pose via Nav2
    and forwards intake / placement / toggle actions to the world services."""

    def __init__(self):
        super().__init__('ai_driver')

        self.nav_client = ActionClient(self, NavigateToPose, 'navigate_to_pose')
        self.marker_pub = self.create_publisher(Marker, '/ai_goal_marker', 10)

        self.create_subscription(
            ActionState, '/sai_output', self.interpret_action, 10)
        self.create_subscription(
            TwistStamped, '/cmd_vel', self.store_current_velocity, 10)

        self.intake_element = self.create_client(
            IntakeElement, '/robot_intake')
        self.score_element = self.create_client(
            ScoreElement, '/score_element')

        self._marker_id_counter = 0

    @staticmethod
    def quaternion_from_euler(roll, pitch, yaw):
        cy = math.cos(yaw * 0.5)
        sy = math.sin(yaw * 0.5)
        cp = math.cos(pitch * 0.5)
        sp = math.sin(pitch * 0.5)
        cr = math.cos(roll * 0.5)
        sr = math.sin(roll * 0.5)
        return [
            cy * cp * cr + sy * sp * sr,
            cy * cp * sr - sy * sp * cr,
            sy * cp * sr + cy * sp * cr,
            sy * cp * cr - cy * sp * sr,
        ]

    def store_current_velocity(self, msg: TwistStamped):
        self.avg_vel = (msg.twist.linear.x + msg.twist.linear.y) / 2

    def publish_goal_marker(self, x, y, robot_color):
        marker = Marker()
        marker.header.frame_id = 'map'
        marker.header.stamp = self.get_clock().now().to_msg()
        marker.id = self._marker_id_counter
        self._marker_id_counter += 1
        marker.type = Marker.SPHERE
        marker.action = Marker.ADD
        marker.pose.position.x = x
        marker.pose.position.y = y
        marker.pose.position.z = 0.1
        marker.pose.orientation.w = 1.0
        marker.scale.x = 0.2
        marker.scale.y = 0.2
        marker.scale.z = 0.2
        marker.lifetime.sec = 99999
        marker.lifetime.nanosec = 0
        if robot_color == 'red':
            marker.color.r = 1.0
        else:
            marker.color.b = 1.0
        marker.color.a = 1.0
        self.marker_pub.publish(marker)

    def drive_to(self, x, y, theta):
        goal_pose = PoseStamped()
        goal_pose.header.frame_id = 'map'
        goal_pose.header.stamp = self.get_clock().now().to_msg()
        goal_pose.pose.position.x = x
        goal_pose.pose.position.y = y
        q = self.quaternion_from_euler(0, 0, theta)
        goal_pose.pose.orientation.x = q[0]
        goal_pose.pose.orientation.y = q[1]
        goal_pose.pose.orientation.z = q[2]
        goal_pose.pose.orientation.w = q[3]
        goal_msg = NavigateToPose.Goal()
        goal_msg.pose = goal_pose
        self.nav_client.send_goal_async(goal_msg)

    def interpret_action(self, msg: ActionState):
        self.publish_goal_marker(msg.red_robot.x, msg.red_robot.y, 'red')
        self.publish_goal_marker(msg.blue_robot.x, msg.blue_robot.y, 'blue')

        if msg.red_robot_action == 0:
            self.drive_to(msg.red_robot.x, msg.red_robot.y, msg.red_robot.theta)
        # action codes 1..4 (intake / place pin / place cup / flip toggle) are
        # handled by the robot's own controller / future manipulator stack.


def main(args=None):
    rclpy.init(args=args)
    node = AIDriver()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/ai_driver.cpp
// ai_driver.cpp - executes the Strategy AI output: drives to the requested
// pose via Nav2 and forwards intake / placement / toggle actions to the world
// services.
#include <chrono>
#include <cmath>
#include <functional>
#include <memory>

#include <geometry_msgs/msg/pose_stamped.hpp>
#include <geometry_msgs/msg/twist_stamped.hpp>
#include <nav2_msgs/action/navigate_to_pose.hpp>
#include <rclcpp/rclcpp.hpp>
#include <rclcpp_action/rclcpp_action.hpp>
#include <visualization_msgs/msg/marker.hpp>

#include "override_sim/msg/action_state.hpp"
#include "override_sim/srv/intake_element.hpp"
#include "override_sim/srv/score_element.hpp"

namespace override_sim
{

using NavigateToPose = nav2_msgs::action::NavigateToPose;

class AIDriver : public rclcpp::Node
{
public:
  AIDriver() : Node("ai_driver")
  {
    nav_client_ = rclcpp_action::create_client<NavigateToPose>(
      this, "navigate_to_pose");
    marker_pub_ = this->create_publisher<visualization_msgs::msg::Marker>(
      "/ai_goal_marker", 10);

    sai_sub_ = this->create_subscription<ActionState>(
      "/sai_output", 10,
      [this](const ActionState::SharedPtr msg) { interpret_action(msg); });
    cmd_vel_sub_ = this->create_subscription<geometry_msgs::msg::TwistStamped>(
      "/cmd_vel", 10,
      [this](const geometry_msgs::msg::TwistStamped::SharedPtr msg) {
        avg_vel_ = (msg->twist.linear.x + msg->twist.linear.y) / 2.0;
      });

    intake_client_ = this->create_client<IntakeElement>("/robot_intake");
    score_client_ = this->create_client<ScoreElement>("/score_element");

    marker_id_counter_ = 0;
  }

private:
  void publish_goal_marker(double x, double y, bool red)
  {
    visualization_msgs::msg::Marker marker;
    marker.header.frame_id = "map";
    marker.header.stamp = this->now();
    marker.id = marker_id_counter_++;
    marker.type = visualization_msgs::msg::Marker::SPHERE;
    marker.action = visualization_msgs::msg::Marker::ADD;
    marker.pose.position.x = x;
    marker.pose.position.y = y;
    marker.pose.position.z = 0.1;
    marker.pose.orientation.w = 1.0;
    marker.scale.x = 0.2;
    marker.scale.y = 0.2;
    marker.scale.z = 0.2;
    marker.lifetime = rclcpp::Duration::from_seconds(99999);
    if (red) {
      marker.color.r = 1.0;
    } else {
      marker.color.b = 1.0;
    }
    marker.color.a = 1.0;
    marker_pub_->publish(marker);
  }

  void drive_to(double x, double y, double theta)
  {
    if (!nav_client_->wait_for_action_server(std::chrono::seconds(1))) {
      RCLCPP_WARN(this->get_logger(), "nav2 action server not available");
      return;
    }
    NavigateToPose::Goal goal;
    goal.pose.header.frame_id = "map";
    goal.pose.header.stamp = this->now();
    goal.pose.pose.position.x = x;
    goal.pose.pose.position.y = y;
    goal.pose.pose.orientation.x = 0.0;
    goal.pose.pose.orientation.y = 0.0;
    goal.pose.pose.orientation.z = std::sin(theta / 2.0);
    goal.pose.pose.orientation.w = std::cos(theta / 2.0);
    nav_client_->async_send_goal(goal);
  }

  void interpret_action(const ActionState::SharedPtr msg)
  {
    publish_goal_marker(msg->red_robot.x, msg->red_robot.y, true);
    publish_goal_marker(msg->blue_robot.x, msg->blue_robot.y, false);

    if (msg->red_robot_action == 0) {
      drive_to(msg->red_robot.x, msg->red_robot.y, msg->red_robot.theta);
    }
    // action codes 1..4 (intake / place pin / place cup / flip toggle) are
    // handled by the robot's own controller / future manipulator stack.
  }

  double avg_vel_ = 0.0;
  int marker_id_counter_ = 0;

  rclcpp_action::Client<NavigateToPose>::SharedPtr nav_client_;
  rclcpp::Publisher<visualization_msgs::msg::Marker>::SharedPtr marker_pub_;
  rclcpp::Subscription<ActionState>::SharedPtr sai_sub_;
  rclcpp::Subscription<geometry_msgs::msg::TwistStamped>::SharedPtr cmd_vel_sub_;
  rclcpp::Client<IntakeElement>::SharedPtr intake_client_;
  rclcpp::Client<ScoreElement>::SharedPtr score_client_;
};

}  // namespace override_sim

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<override_sim::AIDriver>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
